iceoryx_posh/source/popo/user_trigger.cpp

```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2020 by Robert Bosch GmbH, Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "iceoryx_posh/popo/user_trigger.hpp"
+#include "iceoryx_posh/popo/wait_set.hpp"
+
+namespace iox
+{
+namespace popo
+{
+cxx::expected<WaitSetError> UserTrigger::attachToWaitset(WaitSet& waitset,
+                                                         const uint64_t triggerId,
+                                                         const Trigger::Callback<UserTrigger> callback) noexcept
+{
+    m_trigger.reset();
+    m_trigger = Trigger();
+
+    return waitset
+        .acquireTrigger(
+            this, {this, &UserTrigger::hasTriggered}, {this, &UserTrigger::unsetTrigger}, triggerId, callback)
+        .and_then([this](Trigger& trigger) { m_trigger = std::move(trigger); });
+}
+
+void UserTrigger::detachWaitset() noexcept
+{
+    unsetTrigger(m_trigger);
+}
+
+void UserTrigger::trigger() noexcept
+{
+    if (m_trigger)
+    {
+        m_wasTriggered.store(true, std::memory_order_relaxed);
+        m_trigger.trigger();
+    }
```

Comment:
We have to talk about races. 

-  Can `trigger` happen in parallel to `attachWaitset` and `detachWaitset`. I would say yes and therefore we had a [lock in the guard condition](https://github.com/eclipse/iceoryx/blob/master/iceoryx_posh/source/popo/guard_condition.cpp#L24) 

- Can `attachWaitset` and `detachWaitset` happen in parallel to `wait` or `timedWait`. We have to discuss

Depending on the answers your Trigger class is broken like it is

Reply:
This is solved.

1. WaitSet returns a TriggerHandle with acquireTrigger and the UserTrigger stores that triggerHandle
2. Every method of the TriggerHandle is threadsafe, including move ctor/assignment.

Therefore, if a WaitSet detaches the TriggerHandle, the handle is invalidated in a threadsafe way so that a `trigger` call does not interfere/race with the `invalidate` call. Additionally, the `trigger` method checks internally if the handle is valid, if its not valid nothing happens.

If someone attaches a TriggerHandle the move assignment overrides the existing handle in a threadsafe manner so it cannot race with a trigger call of the old handle.